src/server/pps/server/poller.go

```diff
@@ -119,27 +147,133 @@ func (a *apiServer) pollPipelines(pachClient *client.APIClient) {
 		// always rm 'pipeline', to advance loop
 		delete(etcdPipelines, pipeline)
 
-		// generate an etcd event for 'pipeline' by reading it & writing it back
+		// generate a pipeline event for 'pipeline'
 		log.Debugf("PPS master: polling pipeline %q", pipeline)
-		var curPI pps.EtcdPipelineInfo
-		_, err := col.NewSTM(ctx, a.env.GetEtcdClient(), func(stm col.STM) error {
-			return a.pipelines.ReadWrite(stm).Update(pipeline, &curPI,
-				func() error { /* no modification, just r+w */ return nil })
-		})
-		if col.IsErrNotFound(err) {
-			log.Warnf("%q polling conflicted with delete (not found)", pipeline)
-		} else if err != nil {
-			// No sensible error recovery here, just keep polling...
-			log.Errorf("could not poll %q: %v", pipeline, err)
-		}
+		m.eventCh <- &pipelineEvent{eventType: writeEv, pipeline: pipeline}
 
-		// move to next pipeline
+		// move to next pipeline (after 2s sleep)
 		return backoff.ErrContinue
-	}, backoff.NewConstantBackOff(2*time.Second),
+	}, backoff.NewConstantBackOff(pollBackoffTime),
 		backoff.NotifyContinue("pollPipelines"),
 	); err != nil {
 		if ctx.Err() == nil {
 			panic("pollPipelines is exiting prematurely which should not happen; restarting pod...")
 		}
 	}
 }
+
+func (m *ppsMaster) pollPipelinePods(pollClient *client.APIClient) {
+	ctx := pollClient.Ctx()
+	if err := backoff.RetryUntilCancel(ctx, func() error {
+		// watchChan will be nil if the Watch call below errors, this means
+		// that we won't receive events from k8s and won't be able to detect
+		// errors in pods. We could just return that error and retry but that
+		// prevents pachyderm from creating pipelines when there's an issue
+		// talking to k8s.
+		kubePipelineWatch, err := m.a.env.GetKubeClient().CoreV1().Pods(m.a.namespace).Watch(
+			metav1.ListOptions{
+				LabelSelector: metav1.FormatLabelSelector(metav1.SetAsLabelSelector(
+					map[string]string{
+						"component": "worker",
+					})),
+				Watch: true,
+			})
+		if err != nil {
+			return errors.Wrap(err, "failed to watch kubernetes pods")
```

Comment:
not sure what type of errors we expect here - is there any concern about this just constantly looping if there's some k8s connection issue?

Reply:
This is a good question; reflecting on it a little, I *think* this is the right choice. Since this gets run when the PPS master starts, the options I see are:
1. Restart all of pachd (i.e. crash & reboot)
2. Restart the PPS master (or yield the PPS master lock and start the PPS master somewhere else; i.e. return an error)
3. Retry the k8s `Watch()` RPC (what this does)

My rationale for preferring option 3 is this: it seems like the high-level goal is to start a functional k8s watch as quickly as possible. I don't see any reason why a k8s watch is more likely to succeed if we retry it on a different machine. Also, because this code runs on master startup (which happens on pachd startup, and possibly other times as well), I don't think there's any broader context for this code that would need to be part of a retry loop. And shrinking the retry loop to just this function seems a lot faster than the other two options (since we don't have to deal with scheduling or master election), so I think this is the best option.